### PR TITLE
added lastUpdateTime field to status of DNS provider and entry

### LINKED ISSUE #128: Add lastUpdateTime to status fields
**What would you like to be added**:
As an operator I'd like to see a `lastUpdateTime` timestamp for the status sub resources of `DNSEntry`, `DNSOwner`, `DNSProvider`.

Example for `DNSEntry`:
```
status:
  lastUpdateTime: "2020-10-07T00:16:28Z"
  message: dns entry active
  observedGeneration: 1
  provider: provider-name
  providerType: provider-type
  state: Ready
  targets:
  - 8.8.4.4
  ttl: 600
  zone: xyz
```

The `lastUpdateTime` is only supposed to be updated if the DNS record itself is update or the `state` or `message` changes.

**Why is this needed**:
Exposing a `lastUpdateTime` timestamp provides essential information for troubleshooting.

### PR DESCRIPTION
**Pull request by MartinWeindel**
**What this PR does / why we need it**:
The field `lastUpdateTime` is added to the status of `DNSProvider` and `DNSEntry` to give additional information to operators on troubleshooting.

**Which issue(s) this PR fixes**:
Fixes #128 

**Special notes for your reviewer**:

**Release note**:

```improvement operator
added lastUpdateTime field to status of DNSProvider and DNSEntry
```

### CODE REVIEW

**Review by timuthy — CHANGES_REQUESTED** [review]
Thanks for implementing this requirement so quickly. Looks good in general, just some minor suggestions.

File: charts/external-dns-management/templates/crds.yaml [hunk 1]
```diff
@@ -83,6 +83,9 @@ spec:
           type: object
         status:
           properties:
+            lastUpdateTime:
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply charts/external-dns-management/templates/crds.yaml:86]: good catch! done.

File: pkg/apis/dns/v1alpha1/dnsentry.go [hunk 2]
```diff
@@ -82,6 +82,9 @@ type DNSEntryStatus struct {
 	// message describing the reason for the state
 	// +optional
 	Message *string `json:"message,omitempty"`
+	// lastUpdateTime contains the timestamp of the last status update
+	// +optional
+	LastUptimeTime *string `json:"lastUpdateTime,omitempty"`
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/apis/dns/v1alpha1/dnsentry.go]: done

File: pkg/apis/dns/v1alpha1/dnsprovider.go [hunk 3]
```diff
@@ -87,6 +87,9 @@ type DNSProviderStatus struct {
 	State string `json:"state"`
 	// message describing the reason for the actual state of the provider
 	Message *string `json:"message,omitempty"`
+	// lastUpdateTime contains the timestamp of the last status update
+	// +optional
+	LastUptimeTime *string `json:"lastUpdateTime,omitempty"`
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/apis/dns/v1alpha1/dnsprovider.go]: done

File: pkg/dns/utils/utils.go [hunk 4]
```diff
@@ -83,3 +86,7 @@ func (this *LogMessage) Warnf(logger logger.LogContext, add ...interface{}) bool
 func (this *LogMessage) Debugf(logger logger.LogContext, add ...interface{}) bool {
 	return this.out(logger.Debugf, add...)
 }
+
+func SetLastUpdateTime(lastUptimeTime **metav1.Time) {
```
Comment (timuthy) [inline pkg/dns/utils/utils.go:90]: Do you really need `**metav1.Time`?
With `*metav1.Time` you are already capable of manipulating the content of the address space `lastUptimeTime` points to.
Can you also add documentation?

Comment (MartinWeindel) [reply pkg/dns/utils/utils.go:90]: Yes, the double pointer is needed. Initially the `*metav1.Time` is nil and you need to assign the address of the struct.
Regarding documentation: if you call `kubectl get crd dnsentries` you get a description for all status fields including the new field `lastUpdateTime`.

Comment (timuthy) [reply pkg/dns/utils/utils.go:90]: Okay, thanks. I missed the `nil` case.
> Regarding documentation: if you call kubectl get crd dnsentries you get a description for all status fields including the new field lastUpdateTime.
This was related to the function. Having an inline comment for exported elements is considered best practice in Go, [ref](https://golang.org/doc/effective_go.html#commentary).

Comment (MartinWeindel) [reply pkg/dns/utils/utils.go:91]: Ah ok. I also added some documentation for the status of `DNSProvider` and `DNSEntry`. See `/docs/usage`

**Review by timuthy — APPROVED** [review]
/lgtm
Thanks for addressing the comments so quickly.

**Pull request merged**